### PR TITLE
Extremely WIP upgrade to v8

### PR DESCRIPTION
**Pull request by nul800sebastiaan**
This is a very very early exploration of about 2 hours to see how far I could get with upgrading this to v8. I am pretty sure nothing much works. I did get it to render in the backoffice but picking a doctype is where it breaks (see screenshot, that's how far I got so far).

There's absolutely no need to merge this PR, but feel free to use the updated code as examples, which is mostly what I tried this for!

ltjgc_78

### CODE REVIEW

**Comment by ArnoldV** [comment]
Took a look at this and here are some remarks to possibly help someone move further.
My time is unfortunately a bit limited for now to go all the way and make this work.

In the doctypegrideditor.controller.js there is a reference to tabs on line 84:
`                    for (var t = 0; t < newModel.node.tabs.length; t++) {
                        var tab = newModel.node.tabs[t];
                        for (var p = 0; p < tab.properties.length; p++) {
                            var prop = tab.properties[p];
                            if (typeof prop.value !== "function") {
                                value[prop.alias] = prop.value;
                            }
                        }
                    }`

to get access to tabs in 8.0 you'll need to do something like:

`                    for (var v = 0; v < newModel.node.variants.length; v++) {
                            var variant = newModel.node.variants[v];
                        for (var t = 0; t < variant.tabs.length; t++) {
                            var tab = variant.tabs[t];
                            for (var p = 0; p < tab.properties.length; p++) {
                                var prop = tab.properties[p];
                                if (typeof prop.value !== "function") {
                                    value[prop.alias] = prop.value;
                                }
                            }
                        }
                    }`

and again on line 230:
`                        for (var t = 0; t < data.tabs.length; t++) {
                            var tab = data.tabs[t];
                            for (var p = 0; p < tab.properties.length; p++) {
                                var prop = tab.properties[p];
                                if ($scope.model.dialogData.value[prop.alias]) {
                                    prop.value = $scope.model.dialogData.value[prop.alias];
                                }
                            }
                        }`

should now be something like:
`                        for (var v = 0; v < data.variants.length; v++) {
                            var variant = data.variants[v];
                            for (var t = 0; t < variant.tabs.length; t++) {
                                var tab = variant.tabs[t];
                                for (var p = 0; p < tab.properties.length; p++) {
                                    var prop = tab.properties[p];
                                    if ($scope.model.dialogData.value[prop.alias]) {
                                        prop.value = $scope.model.dialogData.value[prop.alias];
                                    }
                                }
                            }
                         }`


But not sure if it it's needed to loop over all the variants for DTGE?

With this change, you seem to be able to select the doctype (click on it) then click 'done'.

Then it errors in the DocTypeGridEdtiorApiController, seems not to work with DI properly there yet.

To fix a few errors:
Line 93: 
`contentTypeAlias = Services.ContentTypeService.GetAliasByGuid(docTypeGuid);`
to:
`contentTypeAlias = Current.Services.ContentTypeService.GetAliasByGuid(docTypeGuid);`

Line 94:
`var contentType = _contentTypeService.Get(contentTypeAlias);`
to:
`var contentType = Current.Services.ContentTypeService.Get(contentTypeAlias);`

Line 194:
`page = _contentCache.GetById(pageId);`
to:
`page = Umbraco.Content(pageId);`

Then there is an error with the culture, but a simple nullcheck (line 177) can fix that.



Furthermore we now have IPublishedElement, which seems to be perfect for this?
So the PreviewModel should be of IPublishedElement instead of IPublishedContent? And remove all the things not needed. Als in DocTypeGridEditorHelper.cs


I got stuck on getting the current page in DocTypeGridEditorHelper.cs
` var pcr = Current.UmbracoContext.PublishedRequest;`

**Comment by nul800sebastiaan** [comment]
I didn't verify any of this 🙈 

But happy to merge so it can move on. I would like someone to take ownership of the PR instead of me though, since I won't have time to complete it. So just copy it, run with it and submit it from your own fork to the main Umco repo!

**Comment by ArnoldV** [comment]
@nul800sebastiaan I can take ownership, but maybe @leekelleher would be the best option?

**Comment by leekelleher** [comment]
> I can take ownership

@ArnoldV Be careful what you wish for! 😃 DTGE for v8 would come with maintenance responsibilities. 😊 

---

While I applaud the efforts to move support for v8 forwards, 👏 I wouldn't expect this to be purely a technical/code-refactoring concern. Other considerations, (off the top of my head), would be...

- Would DTGE take advantage of v8's Infinite Editing capabilities?
- Could it leverage NuCache in any way? _(even I'm not 100% sure what I mean by this)_
- How would **tabs** be displayed? Would it follow UI/UX from the main content editing?
- How would language-variant split-view work? _(how do other grid-editors handle this?)_

I appreciate that some folk want this done ASAP so they can move on with v8, _which is fair enough_, I just want to flag up potential v8 UX considerations too. 🙏

**Comment by ArnoldV** [comment]
@leekelleher ownership of THIS current PR to accept stuff, not of the whole thing ;)

For now it would be something to get this working as is with v7. Then move on from there.

Since you use PublishedContent already, NuCache is a given and would 'just work'.

As is now Umbraco Core with nested content, you mark those doctypes as "Is an Element" type.
That will make them 'simple'. The whole 'grid' in-full can be language variant, but not specific cells. Same for split view etc. 

Now moving on to Grid v2 it can be different, but for Grid 1 version that limitations are very acceptable and on-par with Umbraco 7 version.

**Comment by skttl** [comment]
I'm in for the MVP upgrade where functionality is like in 7. I guess DTGE is going to be obsolete once Grid2 gets done (if it does).

**Comment by leekelleher** [comment]
I have concerns about who will maintain the v8 version of DTGE, (after this PR is complete, merged & released)? That's not to hold up continued work on this PR, I applaud the effort! 👏 

@skttl would you be interested? Knowing that you've previously made excellent contributions to DTGE.

**Comment by skttl** [comment]
@leekelleher Yes, I could be interested. I might be underqualified, for parts of the codebase, but with help from the community it should be possible :)

**Comment by FransdeJong** [comment]
So if I want to collaborate on this how would I start? Can I clone the pull request somehow? I'm not that experienced with git...

**Comment by skttl** [comment]
Hi Frans

I have some work I haven't pushed (not much, but makes the doctype selection work, and shows the fields). I'll push to my fork tomorrow, then you can add that one as a remote and make PRs to the V8 branch.

I'm in a country house for a weekend without my laptop, so can't do it now, sorry 🙂

**Comment by skttl** [comment]
Opened up a new one here:
#137

File: src/Our.Umbraco.DocTypeGridEditor/Extensions/ContentTypeServiceExtensions.cs [hunk 1]
```diff
@@ -9,12 +9,15 @@ internal static class ContentTypeServiceExtensions
...
                 string.Concat("Our.Umbraco.DocTypeGridEditor.Web.Extensions.ContentTypeServiceExtensions.GetAliasById_", id),
-                () => ApplicationContext.Current
-                    .DatabaseContext
-                    .Database
-                    .ExecuteScalar<string>("SELECT [cmsContentType].[alias] FROM [cmsContentType] INNER JOIN [umbracoNode] ON [cmsContentType].[nodeId] = [umbracoNode].[id] WHERE [umbracoNode].[uniqueID] = @0", id));
+                () =>
+                {
+                    using (var scope = Current.ScopeProvider.CreateScope())
+                        return scope.Database.ExecuteScalar<string>("SELECT [cmsContentType].[alias] FROM [cmsContentType] INNER JOIN [umbracoNode] ON [cmsContentType].[nodeId] = [umbracoNode].[id] WHERE [umbracoNode].[uniqueID] = @0", id);
+                });
```
Comment (bjarnef) [inline src/Our.Umbraco.DocTypeGridEditor/Extensions/ContentTypeServiceExtensions.cs]: @nul800sebastiaan I guess this need to call scope.Complete(); or add autoComplete: true in CreateScope() (if it only read data).
https://our.umbraco.com/forum/umbraco-8/96075-alternative-solution-to-use-applicationcontextcurrentdatabasecontextdatabase-in-umbraco-8

Comment (nul800sebastiaan) [reply src/Our.Umbraco.DocTypeGridEditor/Extensions/ContentTypeServiceExtensions.cs]: @bjarnef - You're only reading, so yes: you can set `autoComplete: true` in this case. 👍

**Comment by leekelleher** [comment]
@skttl With PR #137, are you suggesting that we close this PR (#136)?

**Comment by skttl** [comment]
Yes, then we can collab in this PR

**Comment by bjarnef** [comment]
It would be great to use position `center` or `target` for `umb-overlay`, so it is consistent with core overlay to insert grid editors. Furthermore does the editor need to select `doc type grid editor` first as in v7 (not sure is this have been changed to list the defined editors merged with other grid editors).
https://github.com/umco/umbraco-doc-type-grid-editor/blob/36d35c243abb36e9ea91d3175a197fc25a1d9619/src/Our.Umbraco.DocTypeGridEditor/Web/UI/App_Plugins/DocTypeGridEditor/Views/doctypegrideditor.html#L15-L19

image

**Comment by leekelleher** [comment]
Closing this Pull Request (#136), it has been superseded by #137 - thanks @skttl!
@bjarnef You may want to cross-post your latest feedback over on #137?

**Pull request closed without merging**